ambry-frontend/src/main/java/com/github/ambry/frontend/NamedBlobDeleteHandler.java

```diff
@@ -0,0 +1,128 @@
+      throws RestServiceException {
+    accountAndContainerInjector.injectAccountAndContainerForNamedBlob(restRequest,
+        metrics.deleteBlobMetricsGroup);
+    RequestPath requestPath = getRequestPath(restRequest);
+    // Get blob id and reconstruct RequestPath
+    // this is the three-part named blob id, including account name, container name and the custom blob name.
+    String namedBlobId = requestPath.getOperationOrBlobId(false);
+    NamedBlobPath namedBlobPath = NamedBlobPath.parse(namedBlobId, Collections.emptyMap());
+    // This is a three steps operation
+    // 1. Get regular blob id from named blob id.
+    // 2. Delete regular blob id from Router
+    // 3. Delete database row.
+    // If we failed at step 1, or step 1 returns NotFound, then  we just return the error back to client, indicating failure
+    // If we failed at step 2, client can resend the request to ambry so we can resume the operation. If regular blob is already
+    //    deleted, Router would do a nop and return success
+    // If we failed at step 3, client can still resend the delete request to delete this blob again.
+    namedBlobDb.get(namedBlobPath.getAccountName(), namedBlobPath.getContainerName(), namedBlobPath.getBlobName())
+        .thenApply(NamedBlobRecord::getBlobId)
+        .whenComplete((blobId, exception) -> {
+          if (exception != null) {
+            Exception dbException = Utils.extractFutureExceptionCause(exception);
+            if (dbException instanceof RestServiceException
+                && ((RestServiceException) dbException).getErrorCode() == RestServiceErrorCode.Deleted) {
+              // Already deleted from named blob database, return as nop.
+              Exception responseException = null;
+              try {
+                restResponseChannel.setStatus(ResponseStatus.Accepted);
+                restResponseChannel.setHeader(RestUtils.Headers.DATE, new GregorianCalendar().getTime());
+                restResponseChannel.setHeader(RestUtils.Headers.CONTENT_LENGTH, 0);
+              } catch (Exception restResponseException) {
+                responseException = restResponseException;
+              }
+              callback.onCompletion(null, responseException);
+            } else {
+              callback.onCompletion(null, dbException);
+            }
+            return;
+          }
+          RequestPath newRequestPath =
+              new RequestPath(requestPath.getPrefix(), requestPath.getClusterName(), requestPath.getPathAfterPrefixes(),
+                  "/ " + blobId, requestPath.getSubResource(), requestPath.getBlobSegmentIdx());
+          // Replace RequestPath in the RestRequest and call DeleteBlobHandler.handle.
+          restRequest.setArg(InternalKeys.REQUEST_PATH, newRequestPath);
+          try {
+            deleteBlobHandler.handle(restRequest, restResponseChannel, (v, e) -> {
+              if (e != null) {
+                callback.onCompletion(null, e);
```

Comment:
If step 3 fails and user reattempts Delete operation, we may get 404 not found from router. To handle that, would we need to check for 404 here and delete named blob DB row?

Reply:
I suppose we can do that.